hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java

```diff
@@ -224,29 +225,33 @@ public ReferenceCounted<IOmMetadataReader, SnapshotCache> get(String key,
    * @param key snapshot table key
    */
   public void release(String key) {
-    ReferenceCounted<IOmMetadataReader, SnapshotCache> rcOmSnapshot =
-        dbMap.get(key);
-    Preconditions.checkNotNull(rcOmSnapshot,
-        "Key '" + key + "' does not exist in cache");
-
-    if (rcOmSnapshot.decrementRefCount() == 0L) {
-      synchronized (pendingEvictionList) {
-        // Eligible to be closed, add it to the list.
-        pendingEvictionList.add(rcOmSnapshot);
-        // The cache size might have already exceed the soft limit
-        // Thus triggering cleanup() to check and evict if applicable
-        cleanup();
+    dbMap.compute(key, (k, v) -> {
```

Comment:
the cleanup() call below calls dbMap.remove().   can that deadlock with this call to dbMap.compute()?

Reply:
`synchronized` should be reentrant, meaning the same thread that already holds the object intrinsic lock can grab it again. So this should not be an issue in theory.

https://docs.oracle.com/javase/tutorial/essential/concurrency/locksync.html

https://stackoverflow.com/a/249907